File: Semestr3/Paradygmaty-Programowania/Laboratoria/Lab3/zadania.ml

let reverse list = 
  let rec iter (list, accum) = 
    match list with 
      | head :: tail -> iter (tail, head :: accum)
      | []  -> accum
  in iter (list, []);;

let divideList list = 
  let rec iter (list, idx, p1, p2, p3) = 
    match list with 
   | head :: tail ->
           (
            match idx mod 3 with
              | 0 -> iter (tail, idx+1, head :: p1, p2, p3)
              | 1 -> iter (tail, idx+1, p1, head::p2, p3)
              | _ -> iter (tail, idx+1, p1, p2, head :: p3)
           )
   | [] -> (reverse p1 ,reverse p2 , reverse p3)
   in iter (list, 0, [], [], []);;


let t1 = divideList [5;4;3;2];;
let t2 = divideList [1;2;3;4;5;6;7;8;9];;
let t3 = divideList [];;
let t4 = divideList [10];;
let t5 = divideList [1;2;3;4;5;6;7];;

File: Semestr3/Paradygmaty-Programowania/Laboratoria/Lab3/zadania.scala
import scala.annotation.tailrec
object zadania extends App
{
    def reverse [A] (list: List[A]) : List[A] = {
        @tailrec
        def iter (list: List[A], accum: List[A]) : List[A] = {
            list match
                case head :: tail => iter(tail, head :: accum)
                case Nil => accum
        }   
        iter (list, Nil)
    }

    def addVectors (list1 : List[Double], list2: List[Double]) : List[Double] = {
        val neutralValue = 0
        @tailrec
        def iter (list1: List[Double], list2: List[Double], result: List[Double]) : List[Double] = {
            (list1, list2) match
                case (h1::t1, h2::t2) => iter (t1, t2, (h1+h2) :: result)
                case (Nil, h2::t2) => iter (list1, t2, (neutralValue + h2):: result)
                case (h1::t1, Nil) => iter (t1, list2, (neutralValue + h1):: result)
                case (Nil,Nil) => result.reverse
            }
        iter(list1, list2, Nil)
    }

println(addVectors(List(1,2,3), List(4,5,6,7,8)))  
println(addVectors(List(1.3,2.9,3,4.4,5,6), List(1,2,3))) 
println(addVectors(Nil, List(1,2,3)))               
println(addVectors(List(1,2,3), Nil))             
println(addVectors(Nil, Nil))   

} 
